libkatana_python_native/src/PropertyGraph.cpp

```diff
@@ -0,0 +1,742 @@
+#define NPY_NO_DEPRECATED_API NPY_1_11_API_VERSION
+
+#include <numpy/ndarrayobject.h>
+
+// Must come after numpy/ndarrayobject.h since that is required, but
+// not included by arrow/python headers.
+#include <arrow/python/numpy_convert.h>
+#include <arrow/python/numpy_to_arrow.h>
+#include <arrow/python/python_to_arrow.h>
+#include <pybind11/numpy.h>
+#include <pybind11/pybind11.h>
+#include <pybind11/stl.h>
+
+#include "katana/python/Conventions.h"
+#include "katana/python/CythonIntegration.h"
+#include "katana/python/EntityTypeManagerPython.h"
+#include "katana/python/ErrorHandling.h"
+#include "katana/python/NumbaSupport.h"
+#include "katana/python/PropertyGraphPython.h"
+#include "katana/python/PythonModuleInitializers.h"
+#include "katana/python/TemplateSupport.h"
+#include "katana/python/TypeBans.h"
+
+namespace py = pybind11;
+
+katana::Result<std::shared_ptr<arrow::Table>>
+katana::python::PythonArgumentsToTable(
+    const pybind11::object& table, const pybind11::dict& kwargs) {
+  std::shared_ptr<arrow::Table> arrow_table;
+
+  if (arrow::py::is_table(table.ptr())) {
+    arrow_table = KATANA_CHECKED(arrow::py::unwrap_table(table.ptr()));
+  } else if (!table.is_none()) {
+    arrow_table = KATANA_CHECKED(arrow::py::unwrap_table(
+        py::module::import("pyarrow").attr("table")(table).ptr()));
+  }
+
+  for (const auto& [k, v] : kwargs) {
+    std::shared_ptr<arrow::ChunkedArray> array;
+    if (py::isinstance<py::array>(v)) {
+      // Convert a numpy array.
+      py::array pyarray = py::cast<py::array>(v);
+      std::shared_ptr<arrow::DataType> dtype;
+      KATANA_CHECKED(
+          arrow::py::NumPyDtypeToArrow(pyarray.dtype().ptr(), &dtype));
+      KATANA_CHECKED(arrow::py::NdarrayToArrow(
+          arrow::default_memory_pool(), v.ptr(), nullptr, true, dtype,
+          arrow::compute::CastOptions::Safe(dtype), &array));
+      KATANA_LOG_DEBUG_ASSERT(array);
+    } else {
+      // Convert any other python sequence.
+      arrow::py::PyConversionOptions options;
+      options.from_pandas = true;
```

Comment:
This is super cool. I wonder if we can do something similar with Dask. It should be possible, but it may just require some trickery.

Reply:
You mean access dask native code from our native code? IT might be possible, I haven't looked. I think dask is almost entirely python though. Not much actual native code other than numpy.

Reply:
Not Dask native code, but maybe construct arrow objects by getting the underlying pandas partitions and trying to create arrow tables from there. Something along those lines.

Reply:
Ah. Yea. We might be able to do that. But it might be easier to get numpy versions of the underlying partitions and work from there. We'll see. Either is possible. pybind11 has rather nice support for numpy.